multicluster/controllers/multicluster/leader/member_cleanup_controller.go

```diff
@@ -0,0 +1,184 @@
+	Scheme    *runtime.Scheme
+	namespace string
+}
+
+func NewMemberResourcesCleanupController(Client client.Client,
+	Scheme *runtime.Scheme,
+	namespace string) *MemberResourcesCleanupController {
+	controller := MemberResourcesCleanupController{
+		Client:    Client,
+		Scheme:    Scheme,
+		namespace: namespace,
+	}
+	return &controller
+}
+
+func (c *MemberResourcesCleanupController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	memberAnnounce := &mcv1alpha1.MemberClusterAnnounce{}
+	err := c.Get(ctx, req.NamespacedName, memberAnnounce)
+	if err != nil {
+		// Clean up all corresponding ResourceExports when the member cluster's MemberClusterAnnounce is deleted.
+		if apierrors.IsNotFound(err) {
+			clusterID := getClusterIDFromName(req.Name)
+			allResExports, err := c.getAllResourceExports(ctx)
+			if err != nil {
+				klog.V(2).ErrorS(err, "Failed to get ResourceExports in the Namespace", "namespace", c.namespace)
+				return ctrl.Result{}, err
+			}
+			cleanupSucceed := c.cleanupResources(ctx, clusterID, allResExports)
+			if !cleanupSucceed {
+				return ctrl.Result{}, fmt.Errorf("failed to clean up all stale ResourceExports for the member cluster %s, retry later", clusterID)
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (c *MemberResourcesCleanupController) getAllResourceExports(ctx context.Context) ([]mcv1alpha1.ResourceExport, error) {
+	resourceExports := &mcv1alpha1.ResourceExportList{}
+	err := c.Client.List(ctx, resourceExports, &client.ListOptions{Namespace: c.namespace})
+	if err != nil {
+		return nil, err
+	}
+	return resourceExports.Items, nil
+}
+
+func (c *MemberResourcesCleanupController) checkAndCleanup(ctx context.Context) {
+	validMemberClusterIDs, err := c.getValidMemberClusterIDs(ctx)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get valid member cluster's ClusterID in the Namespace", "namespace", c.namespace)
+		return
+	}
+
+	allResExports, err := c.getAllResourceExports(ctx)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports in the Namespace", "namespace", c.namespace)
+		return
+	}
+
+	allMemberClusterIDs, err := c.getAllMemberClusterIDs(ctx, allResExports)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get all member cluster's ClusterID in the Namespace", "namespace", c.namespace)
+		return
+	}
+
+	staleMemberClusterIDs := allMemberClusterIDs.Difference(validMemberClusterIDs)
+	for clusterID := range staleMemberClusterIDs {
+		cleanupSucceed := c.cleanupResources(ctx, clusterID, allResExports)
+		if cleanupSucceed {
+			klog.InfoS("The member cluster's stale resources are cleaned up", "clusterID", clusterID)
+		} else {
+			klog.Error("Failed to clean up the member cluster's stale resources", "clusterID", clusterID)
+		}
+	}
+}
+
+func (c *MemberResourcesCleanupController) getAllMemberClusterIDs(ctx context.Context, resourceExports []mcv1alpha1.ResourceExport) (sets.Set[string], error) {
```

Comment:
I feel we should directly filter ClusterIDs of ResourceExports and get a list of Exports to delete, rather than compute the ClusterID diff and call cleanupResources() for each non-existing ClusterID, which adds unnecessary complexity.

Reply:
Refined this part.